loader/0002-quilt.mod.json.md

```diff
@@ -0,0 +1,419 @@
+
+### The `environment` field
+| Type   | Required |
+|--------|----------|
+| String | False    |
+
+Defines the environment(s) that this mod should be loaded on. Valid values are:
+* `"*"` — All environments (default)
+* `"client"` — The physical client
+* `"server"` — The dedicated server
+
+## Custom Elements
+In addition to the defined elements above, mods and libraries will be able to add their own elements to the quilt mod file. Mods will be expected to define up to one top-level element corresponding to their mod id. The element can be of any type, so that mods can define either a single value, array of values, or a sub-object.
+
+## Example
+An example quilt.mod.json:
+```json
+{
+    "schema_version": 1,
+    "quilt_loader": {
+        "group_id": "org.quiltmc",
+        "mod_id": "example_mod",
+        "version": "1.0.0",
+        "entrypoints": {
+            "main": [
+                "org.quiltmc.example_mod.impl.ExampleMod",
+                "org.quiltmc.example_mod.impl.ExampleModNetworking"
+            ],
+            // Since we only have a single client endpoint, no array is needed.
+            "client": "org.quiltmc.example_mod.impl.client.ExampleModClient",
+        },
+        "depends": [
+            "quilt_networking_api",
+            "quilt_rendering_api",
+            {
+                "id": "modmenu",
+                "environment": "client"
+            }
+        ],
+        "breaks": [
+            {
+                "id": "sodium",
+                "versions": "*",
+                "reason": "Sodium does not implement the Quilt Rendering API."
+            },
+            {
+                "id": "some_random_library",
+                "versions": [
+                    "1.23.456", // A reason is not required
+                    {
+                        "versions": "<1.0.0",
+                        "reason": "Stable API required"
+                    },
+                    {
+                        "versions": "1.5.3",
+                        "reason": "Contains a game-breaking bug"
+                    }
+                ]
+            }
+        ],
+        "metadata": {
+            "name": "Quilt Example Mod",
+            "description": "An example mod for the Quilt ecosystem.",
+            "contributors": {
+                "Haven King": "Developer"
+            },
+            "contact": {
+                "homepage": "https://quiltmc.org/"
+            },
+            "license": "CC0-1.0",
+            "icon": "assets/modid/icon.png"
+        }
+    },
+    "mixins": [
+        "modid.mixins.json"
+    ]
+}
+```
+
+## Placeholders
+Strings with defined formats such as [the version field](#the-version-field) can instead supply a string matching the regex `^\$\{[a-zA-Z_$][a-zA-Z\d_$]*\}$` in a development environment to be replaced on build. The quilt-gradle plugin defines usage of the value `${version}` as a placeholder for the version declared in gradle.properties.
```

Comment:
Can't `^\$\{[a-zA-Z_$][a-zA-Z\d_$]*\}$` be replaced with `^\$\{[a-zA-Z_$]+\}$` or the `+` not a standard regex?

Reply:
I think this refers to a placeholder that gradle can replace at compliation time such as `{$version}`

Reply:
Those aren't the same character sets, although I didn't know you could use `\d` inside `[]`. I usually just use `0-9` like a normal person.

Reply:
> aren't the same character sets
Ah, sorry, my bad pretty hard to compare text on the phone, when it's a line break (because of scale)

Reply:
It can be simplified to `^\$\{[a-zA-Z_$][\w$]*\}$` tho

Reply:
I believe the current format is somewhat clearer to those not familiar with the intricacies of regular expressions.